doc/specs/#7335 - Console Allocation Policy.md

```diff
@@ -0,0 +1,348 @@
+
+### New APIs
+
+Because a console-subsystem application may still want fine-grained control over when and how its console window is
+spawned, we propose the inclusion of a new API, `AllocConsoleEx(PCONSOLE_ALLOCATE_INFO)`.
+
+#### `AllocConsoleEx`
+
+```c++
+typedef
+enum _CONSOLE_ALLOCATE_MODE
+{
+    CONSOLE_ALLOCATE_DEFAULT = 0,
+    CONSOLE_ALLOCATE_HIDDEN,
+    CONSOLE_ALLOCATE_MINIMIZED
+} CONSOLE_ALLOCATE_MODE;
+
+typedef
+struct _CONSOLE_ALLOCATE_INFO
+{
+    DWORD cbSize;
+    CONSOLE_ALLOCATE_MODE mode;
+} CONSOLE_ALLOCATE_INFO, *PCONSOLE_ALLOCATE_INFO;
+
+WINBASEAPI
+HRESULT
+WINAPI
+AllocConsoleEx(PCONSOLE_ALLOCATE_INFO pAllocateInfo);
+```
+
+**AllocConsoleEx** affords an application control over how its console windows are allocated.
+
+Modes:
+
+* `CONSOLE_ALLOCATE_DEFAULT`: Equivalent to `AllocConsole()` or `AllocConsoleEx(NULL)`. The console window is configured
+  based on the fields in the `STARTUPINFO` used to spawn the current process.
+* `CONSOLE_ALLOCATE_HIDDEN`: Allocates a console with no visible window. The window can be displayed at a later time.
+* `CONSOLE_ALLOCATE_MINIMIZED`: Allocates a console with a window that is minimized to the taskbar.
+
+Applications seeking backwards compatibility are encouraged to delay-load `AllocConsoleEx` or check for its presence in
+the `api-ms-win-core-console-l1` APISet.
+
+## Inspiration
+
+Fusion manifest entries are used to make application-scoped decisions like this all the time, like `longPathAware` and
+`heapType`.
+
+CUI applications that can spawn a UI (or GUI applications that can print to a console) are commonplace on other
+platforms because there is no subsystem differentiation.
+
+## UI/UX Design
+
+There is no UI for this feature.
+
+## Capabilities
+
+### Accessibility
+
+This should have no impact on accessibility.
+
+### Security
+
+One reviewer brought up the potential for a malicious actor to spawn an endless stream of headless daemon processes.
+
+This proposal in no way changes the facilities available to malicious people for causing harm: they could have simply
+used `IMAGE_SUBSYSTEM_WINDOWS_GUI` and not presented a UI--an option that has been available to them for 35 years.
+
+### Reliability
+
+This should have no impact on reliability.
+
+### Compatibility
+
+An existing application opting into **hidden** may constitute a breaking change, but the scope of the breakage is
+restricted to that application and is expected to be managed by the application.
+
+All behavioral changes are opt-in.
+
+> **EXAMPLE**: If Python updates python.exe to specify an allocation policy of **hidden**, graphical python applications
+> will become double-click runnable from the graphical shell without spawning a console window. _However_, console-based
+> python applications will no longer spawn a console window when double-clicked from the graphical shell.
+>
+> If python.exe specifies **detached**, Console APIs will fail until a console is allocated.

```

Comment:
Here are 2 mentions of "hidden". I think the former one also refers to policies, right?

Reply:
Ah, there they are. Thank you!